runtime/src/main/java/org/corfudb/runtime/collections/StreamPollingTask.java

```diff
@@ -0,0 +1,108 @@
+import java.util.concurrent.TimeUnit;
+
+/**
+ * A runnable task for a subscription to poll data change from the
+ * transaction stream and put into the subscription's buffer. This
+ * task is executed by the thread pool continuously until an error
+ * occurs which would stop the subscription.
+ * <p>
+ * Created by WenbinZhu on 11/9/20.
+ */
+@Slf4j
+class StreamPollingTask implements Runnable {
+
+    // A period of time in ms to sleep before next cycle when poller gets no new data changes.
+    private static final Duration IDLE_WAIT_TIME_MS = Duration.ofMillis(50);
+
+    // Total amount of time to wait for putting the polled data changes into buffer if it is full.
+    private static final Duration QUEUE_FULL_BLOCK_TIME_MS = Duration.ofMillis(2_000);
+
+    // The streaming manager that is in charge of listener subscriptions.
+    private final StreamingManager streamingManager;
+
+    // The corfu transaction stream to poll data changes from.
+    private final IStreamView txnStream;
+
+    // The subscription context associated with this task.
+    private final StreamSubscription subscription;
+
+    // The Thread pool for executing stream polling tasks.
+    private final ExecutorService pollingExecutor;
+
+    // Last address of the data successfully processed by the buffer.
+    private long lastReadAddress;
+
+    StreamPollingTask(StreamingManager streamingManager, long lastAddress,
+                      StreamSubscription subscription, ExecutorService executor) {
+        this.streamingManager = streamingManager;
+        this.subscription = subscription;
+        this.pollingExecutor = executor;
+        this.lastReadAddress = lastAddress;
+        this.txnStream = subscription.getTxnStream();
+    }
+
+    @Override
+    public void run() {
+        try {
+            pollTxnStream();
+        } catch (Throwable throwable) {
+            StreamListener listener = subscription.getListener();
+            log.error("Encountered exception {} during txn stream polling, listener: {}, " +
+                    "namespace: {}", throwable, listener, subscription.getNamespace());
+            streamingManager.unsubscribe(listener, false);
+            listener.onError(throwable);
+        }
+    }
+
+    /**
+     * Poll new data changes from the transaction stream and put into
+     * the subscription's buffer one by one.
+     */
+    private void pollTxnStream() throws Exception {
+        // If listener already unsubscribed, do not process or schedule again.
+        if (subscription.isStopped()) {
+            return;
+        }
+
+        // Seek to next address and poll transaction updates.
+        txnStream.seek(lastReadAddress + 1L);
+        List<ILogData> updates = txnStream.remainingAtMost(subscription.getStreamBufferSize());
+
+        // No new updates, take a short break and poll again.
+        if (updates.isEmpty()) {
+            TimeUnit.MILLISECONDS.sleep(IDLE_WAIT_TIME_MS.toMillis());
```

Comment:
This is not the only way to prevent busy polling and can cause high latencies.

If you have 1 thread running the notification tasks and 8 subscriptions and only one that is producing data then, the listener of that active stream will take 7 * 50 milliseconds to deliver the notification. You can try to hide this by increasing the number of threads in the notification task, but then the pathological case can be really bad and it doesn't seem like its unlikely, or rather, its not clear that at most a 50ms penalty will be inured. 

Depending on the producer rate it seems really hard to reason about the worst case. It's better to just use a scheduled executor and push that sleep to the executor scheduling instead of doing it before the submit.

Reply:
I think the calculation is not right here. It's won't wait for 7 * 50 ms. As long as one thread is available, it can pick up the next task. If you have 7 subscriptions which are not producing data, and you have 4 threads in total, the latency of the  next subscription is just 50 ms, because all threads wait in parallel, not sequentially. And if you have 8 threads, there will be no latecy at all. Actually I chose this model because the schduled exector has an intrinsic latency, which is the scheduling interval. This interval causes latecy on every poll, no matter there is data or not, while this model has no latency when there are data produced, which is the common case. Sundar tried to reduce the schduled exector's latecy by using an infinite loop in the task, but that cause scalability issues as mentioned in my Conflucence deisign doc. So I intentionally abandoned the scheduled executor when making the design decision.

Reply:
The max blocking time for a stream that has unread data is a function of the number of threads and number of streams.

If you have 1 threads running the notification task and 7 subscribers with only 1 active stream then the executor queue will look like this | Active | Inactive | Inactive | Inactive | Inactive | Inactive | Inactive (processing thread). If the queue will be consumed from right to left, then the processing thread will sleep for 6 * 50ms before the active thread is discovered. If you use a scheduled executes then the tasks will incur the sleep while waiting in the min-heap of the executor, notice that cost will not be incurred on the processing thread. This means that it can pick up active streams that can be read right away. 

The fundamental difference is do you incur the cost of the sleep on the thread (which stalls the thread), or inside a waiting queue (which allows the worker thread to do useful work for active threads). 

Is that a correct characterization of a possible pattern ?

Reply:
This is a pathological case where your subscriber number is 8x than your polling thread, and most of the polled data is empty. I think the optimization is not related to using a scheduled executor or not, because scheduled executor has an intrintic delay which is the schedule interval. The right optimization, as you and Medhavi both kind of mentioned, is to use a customed delay queue for the executor and we still let the thread itself submit its tasks after the previous one is done, and the tasks will be a DelayedTask in case of empty data. However for now I'm not sure if that is worthy because our workloads are not large enough to have these pathological cases.

Reply:
> However for now I'm not sure if that is worthy because our workloads are not large enough to have these pathological cases.

The max blocking time for a stream that has unread data is a function of the number of threads and number of streams. Therefore in order to determine the likelihood of the worst case happening we need to know the numbers. We currently have ~300 tables in the system (as of ~6 months ago), if we assume that 20% of the tables will be tagged then each thread will on average handle (300 * .2) / 4 = 15 streams. Since most of the data is config data (i.e. static) then its clearly not a pathological case. In the future, we can only assume that there will be more tables.

> executor has an intrintic delay which is the schedule interval

I don't understand where this intrinsic delay comes from and where it is incurred. Please explain. 

I also don't see this as an optimization per-se it just seems like a natural polling strategy, that is, to let the task sleep in the queue and not block the worker.

Reply:
Sure I will explore the delayedQueue approach, but let me clarify the calculation: the blocking time is not related to number of streams, because for each tag we have only one poller. Second, the blocking time is strongly related to number of subscribers, if number of subscribers is less than or equal to the thread number (which is the common case), than there will be no delay at all. An example is If you have 8 subscribers and 4 threads, the maximum possible delay if every poller gets empty data is 8/4*50 = 100ms, not 8 * 50. The reason I'm saying it's an optimization is that our current workload is not even saturating the thread number, i.e. number of subscribers <= thread number.

> I don't understand where this intrinsic delay comes from and where it is incurred. Please explain.

If the scheduled executors you mean is the scheduledExecutorService, than I mean it has an scheduling interval, which is the intrinsic delay, that will be casted on the poller every time, not matter it gets data or not. You can set that interval to be 0, then it's no different than the current approach. I think what you mean is not related to scheduled executor, rather it is the queue inside the executor, that needs to be delayed queue, and that's when the min-heap come into play.

Reply:
> Sure I will explore the delayedQueue approach, but let me clarify the calculation

Yep. sorry my point was Subscriber/tag vs Notification threads. 

> 8/4*50 = 100ms, not 8 * 50.

We are talking about two different configs. 4 threads are currently being used, but I used 1 thread in my example to highlight the relation between notification threads and tags

>If the scheduled executors you mean is the scheduledExecutorService,

What I was referring to was [ScheduledExecutorService::schedule](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ScheduledExecutorService.html#schedule(java.lang.Runnable,%20long,%20java.util.concurrent.TimeUnit)) which creates a one-shot execution that will run after a certain delay. I'm not saying that you should schedule the tasks periodically with same delay. Using the current scheme depending on whether you read yields some entries or not, you can call `schedule(task, 0)` or `schedule(task, Delay)`. This way the task delay will be incurred in the queue waiting and not on the worker thread. `ScheduledExecutorService` internally maintains a heap ordered by the tasks delay time. Now if all the streams are active then most of your schedules will be submitted with a 0 delay and that seems to be be fine when it comes to fairness. Based on code inspection it seems like they will be in FIFO order.

Reply:
> What I was referring to was ScheduledExecutorService::schedule which creates a one-shot execution that will run after a certain delay.

Ok then we are talking about the same idea, but slightly different implementation.

Reply:
It seems like you ended up using the same method. Just out of curiosity, what was the different implementation that you were referring to, explicitly managing the scheduler's min-heap?

Reply:
Yes, I used ScheduledExecutorService. The alternative is still using a normal executor, but give it a customed delayed queue. Initially I thought DelayedWorkQueue is a public utility class that I can use, but later I found it's internal to ScheduledThreadPoolExecutor.